File: cpp/src/broker/paper_broker.cpp
#include <memory>
#include <algorithm>
/**
 * AlgoForge — src/broker/paper_broker.cpp
 * Paper trading broker. Simulates order fills, SL/TP, slippage.
 * Generates synthetic OHLCV price data from a random walk.
 */
#include "broker/broker.hpp"
#include "core/config.hpp"
#include <unordered_map>
#include <vector>
#include <mutex>
#include <cstring>
#include <cmath>
#include <cstdio>
#include <cassert>
#include <random>
#include <algorithm>

namespace af {

/* ── Default spreads ── */
static double default_spread(const char *sym) {
    if (!sym) return 0.0001;
    std::string s(sym);
    if (s=="XAUUSD"||s=="XAGUSD") return 0.25;
    if (s=="USDJPY"||s=="EURJPY") return 0.030;
    if (s.find("JPY")!=std::string::npos) return 0.030;
    return 0.00012;
}

/* ── Price simulator ── */
struct PriceSim {
    double mid = 1.0;
    double volatility = 0.008;
    std::mt19937 rng{42};

    PriceSim() = default;
    PriceSim(double start, double vol, uint32_t seed=42)
        : mid(start), volatility(vol), rng(seed) {}

    double next_mid() {
        std::normal_distribution<double> nd(0.0, volatility / std::sqrt(24.0*60.0));
        mid = std::max(mid * (1.0 + nd(rng)), 0.0001);
        return mid;
    }
};

/* ── PaperBroker ── */
class PaperBroker final : public IBroker {
public:
    explicit PaperBroker(double initial_balance = 10000.0)
        : balance_(initial_balance), equity_(initial_balance)
    {
        /* Seed price simulators */
        price_sim_["EURUSD"] = PriceSim(1.08500, 0.0080, 1);
        price_sim_["GBPUSD"] = PriceSim(1.26500, 0.0100, 2);
        price_sim_["USDJPY"] = PriceSim(149.500, 0.0070, 3);
        price_sim_["XAUUSD"] = PriceSim(2050.00, 0.0120, 4);
        price_sim_["EURJPY"] = PriceSim(162.000, 0.0090, 5);
        price_sim_["GBPJPY"] = PriceSim(188.000, 0.0110, 6);
        price_sim_["AUDUSD"] = PriceSim(0.65500, 0.0090, 7);
        price_sim_["USDCAD"] = PriceSim(1.36000, 0.0075, 8);
        price_sim_["USDCHF"] = PriceSim(0.90500, 0.0070, 9);
        price_sim_["EURGBP"] = PriceSim(0.85800, 0.0060,10);
    }

    AF_Error connect() override {
        connected_ = true;
        snprintf(acct_.currency, sizeof(acct_.currency), "USD");
        acct_.balance     = balance_;
        acct_.equity      = equity_;
        acct_.free_margin = balance_;
        acct_.margin      = 0;
        acct_.leverage    = 100;
        acct_.login       = 99999999;
        printf("[PAPER] Connected | balance=%.2f\n", balance_);
        return AF_OK;
    }

    void disconnect() override { connected_ = false; }
    bool is_connected() const override { return connected_; }
    const char* broker_name() const override { return "PaperSimulator"; }

    AF_Error get_account(AF_AccountInfo &out) const override {
        std::lock_guard lk(mu_);
        update_equity();
        out         = acct_;
        out.balance = balance_;
        out.equity  = equity_;
        return AF_OK;
    }

    AF_Error get_tick(const char *symbol, AF_Tick &out) const override {
        if (!symbol) return AF_ERR_INVALID_SYMBOL;
        std::lock_guard lk(mu_);
        auto it = price_sim_.find(symbol);
        if (it == price_sim_.end()) return AF_ERR_INVALID_SYMBOL;
        double mid = const_cast<PriceSim&>(it->second).next_mid();
        double sp  = default_spread(symbol);
        out.bid    = mid - sp/2.0;
        out.ask    = mid + sp/2.0;
        out.volume = 500.0 + std::uniform_real_distribution<>(0,2000)(
                     const_cast<std::mt19937&>(rng_global_));
        out.timestamp = 0; /* would be time(NULL) in production */
        snprintf(out.symbol, sizeof(out.symbol), "%s", symbol);
        last_ticks_[symbol] = out;
        return AF_OK;
    }

    AF_Error get_symbol_info(const char *symbol, AF_SymbolInfo &out) const override {
        if (!symbol) return AF_ERR_INVALID_SYMBOL;
        snprintf(out.name, sizeof(out.name), "%s", symbol);
        std::string s(symbol);
        out.digits        = (s=="XAUUSD"||s=="XAGUSD") ? 2 : (s.find("JPY")!=std::string::npos?3:5);
        out.point         = std::pow(10.0, -out.digits);
        out.contract_size = 100000.0;
        out.volume_min    = 0.01;
        out.volume_max    = 100.0;
        out.volume_step   = 0.01;
        return AF_OK;
    }

    AF_Error get_bars(const char *symbol, AF_Timeframe tf,
                       int count, AF_Bar *bars_out, int *filled) const override {
        if (!symbol || !bars_out || !filled || count<=0) return AF_ERR_INVALID_PARAM;
        std::lock_guard lk(mu_);
        auto it = price_sim_.find(symbol);
        if (it == price_sim_.end()) return AF_ERR_INVALID_SYMBOL;

        /* Generate synthetic bars via random walk */
        PriceSim sim = it->second;
        double vol_per_bar = 0.0080 / std::sqrt(86400.0 / tf) / std::sqrt(252.0);
        std::mt19937 rng(42 + static_cast<uint32_t>(tf));
        std::normal_distribution<double> nd(0.0, vol_per_bar);

        double price = sim.mid;
        for (int i = 0; i < count; i++) {
            double o = price;
            double h = o, l = o;
            for (int t=0;t<4;t++){
                double p=o*(1+nd(rng));
                if (p>h)h=p; if (p<l)l=p;
            }
            double c = o * (1+nd(rng));
            if (c>h)h=c; if (c<l)l=c;
            bars_out[i].timestamp = 0;
            bars_out[i].open = o; bars_out[i].high = h;
            bars_out[i].low  = l; bars_out[i].close= c;
            bars_out[i].volume= 500+std::uniform_real_distribution<>(0,4500)(rng);
            bars_out[i].spread= default_spread(symbol);
            af_bar_init(&bars_out[i]);
            price = c;
        }
        *filled = count;
        return AF_OK;
    }

    AF_Error place_order(const char *symbol, AF_OrderType type,
                          double lots, double price,
                          double sl, double tp,
                          uint32_t magic, const char *comment,
                          AF_Order &out) override {
        std::lock_guard lk(mu_);
        if (!symbol) return AF_ERR_INVALID_SYMBOL;
        if (lots < 0.01) return AF_ERR_INVALID_LOTS;

        /* Resolve market price */
        AF_Tick tick{};
        auto it = price_sim_.find(symbol);
        if (it == price_sim_.end()) return AF_ERR_INVALID_SYMBOL;
        double mid = const_cast<PriceSim&>(it->second).next_mid();
        double sp  = default_spread(symbol);
        tick.bid = mid-sp/2; tick.ask = mid+sp/2;

        bool is_market = (type==AF_ORDER_BUY||type==AF_ORDER_SELL);
        double fill_p  = 0;
        if (type==AF_ORDER_BUY  || type==AF_ORDER_BUY_STOP)   fill_p = tick.ask;
        if (type==AF_ORDER_SELL || type==AF_ORDER_SELL_STOP)   fill_p = tick.bid;
        if (type==AF_ORDER_BUY_LIMIT)  fill_p = std::min(price, tick.ask);
        if (type==AF_ORDER_SELL_LIMIT) fill_p = std::max(price, tick.bid);

        double commission = 7.0 * lots;

        AF_Position pos{};
        pos.ticket      = next_ticket_++;
        snprintf(pos.symbol, sizeof(pos.symbol), "%s", symbol);
        pos.side        = (type==AF_ORDER_BUY||type==AF_ORDER_BUY_LIMIT||type==AF_ORDER_BUY_STOP)
                          ? AF_DIR_LONG : AF_DIR_SHORT;
        pos.lots        = lots;
        pos.open_price  = fill_p;
        pos.current_price= fill_p;
        pos.sl          = sl;
        pos.tp          = tp;
        pos.profit      = 0;
        pos.commission  = -commission;
        pos.magic       = magic;
        if (comment) snprintf(pos.comment, sizeof(pos.comment), "%s", comment);

        positions_[pos.ticket] = pos;
        balance_ -= commission;

        /* Fill result */
        out.ticket     = pos.ticket;
        out.fill_price = fill_p;
        snprintf(out.symbol, sizeof(out.symbol), "%s", symbol);
        out.type       = type;
        out.lots       = lots;
        out.sl         = sl;
        out.tp         = tp;
        out.magic      = magic;

        printf("[PAPER] %s %.2f %s @ %.5f | SL=%.5f TP=%.5f | comm=-%.2f\n",
               pos.side==AF_DIR_LONG?"BUY":"SELL", lots, symbol, fill_p, sl, tp, commission);

        return AF_OK;
    }

    AF_Error close_position(uint64_t ticket, double lots_to_close = 0.0) override {
        std::lock_guard lk(mu_);
        auto it = positions_.find(ticket);
        if (it == positions_.end()) return AF_ERR_POSITION_NOT_FOUND;
        AF_Position &pos = it->second;

        auto pit = price_sim_.find(pos.symbol);
        double mid = (pit!=price_sim_.end())
                     ? const_cast<PriceSim&>(pit->second).next_mid() : pos.open_price;
        double sp  = default_spread(pos.symbol);
        double exit_price = (pos.side==AF_DIR_LONG) ? mid-sp/2 : mid+sp/2;

        double close_lots = (lots_to_close>0 && lots_to_close < pos.lots)
                            ? lots_to_close : pos.lots;
        double raw_pnl = pos.side==AF_DIR_LONG
                         ? (exit_price - pos.open_price) * close_lots * 100000.0
                         : (pos.open_price - exit_price) * close_lots * 100000.0;
        double comm = 0.0;  /* already charged on open */
        double net  = raw_pnl + comm;

        balance_ += net;
        trade_log_.push_back(pos);
        trade_log_.back().profit = net;
        trade_log_.back().current_price = exit_price;

        printf("[PAPER] CLOSE %s %s %.2f @ %.5f | P&L=%.2f\n",
               pos.symbol, pos.side==AF_DIR_LONG?"LONG":"SHORT",
               close_lots, exit_price, net);

        if (close_lots >= pos.lots) {
            positions_.erase(it);
        } else {
            pos.lots -= close_lots;
        }
        return AF_OK;
    }

    AF_Error modify_position(uint64_t ticket, double sl, double tp) override {
        std::lock_guard lk(mu_);
        auto it = positions_.find(ticket);
        if (it==positions_.end()) return AF_ERR_POSITION_NOT_FOUND;
        it->second.sl = sl;
        it->second.tp = tp;
        return AF_OK;
    }

    AF_Error cancel_order(uint64_t ticket) override {
        std::lock_guard lk(mu_);
        pending_.erase(ticket);
        return AF_OK;
    }

    std::vector<AF_Position> get_positions() const override {
        std::lock_guard lk(mu_);
        std::vector<AF_Position> v;
        for (auto &[t,p]:positions_) { update_position(const_cast<AF_Position&>(p)); v.push_back(p); }
        return v;
    }

    std::vector<AF_Order> get_orders() const override {
        std::lock_guard lk(mu_);
        std::vector<AF_Order> v;
        for (auto &[t,o]:pending_) v.push_back(o);
        return v;
    }

    std::optional<AF_Position> get_position(uint64_t ticket) const override {
        std::lock_guard lk(mu_);
        auto it = positions_.find(ticket);
        if (it==positions_.end()) return std::nullopt;
        return it->second;
    }

    void poll_sl_tp() override {
        std::lock_guard lk(mu_);
        std::vector<uint64_t> to_close;
        for (auto &[ticket, pos] : positions_) {
            auto pit = price_sim_.find(pos.symbol);
            if (pit==price_sim_.end()) continue;
            double mid= const_cast<PriceSim&>(pit->second).next_mid();
            double sp = default_spread(pos.symbol);
            double bid= mid-sp/2, ask= mid+sp/2;
            pos.current_price = (pos.side==AF_DIR_LONG) ? bid : ask;
            if (pos.sl > 0) {
                bool sl_hit = (pos.side==AF_DIR_LONG && bid<=pos.sl) ||
                              (pos.side==AF_DIR_SHORT && ask>=pos.sl);
                if (sl_hit) { to_close.push_back(ticket); continue; }
            }
            if (pos.tp > 0) {
                bool tp_hit = (pos.side==AF_DIR_LONG && bid>=pos.tp) ||
                              (pos.side==AF_DIR_SHORT && ask<=pos.tp);
                if (tp_hit) to_close.push_back(ticket);
            }
        }
        /* Close outside lock — recursive close_position would deadlock */
        mu_.unlock();
        for (uint64_t t : to_close) close_position(t);
        mu_.lock();
    }

    const std::vector<AF_Position>& trade_log() const { return trade_log_; }

private:
    mutable std::mutex mu_;
    bool connected_{false};
    double balance_; mutable double equity_;
    mutable AF_AccountInfo acct_{};

    mutable std::unordered_map<std::string, PriceSim> price_sim_;
    mutable std::unordered_map<std::string, AF_Tick>  last_ticks_;
    std::unordered_map<uint64_t, AF_Position>         positions_;
    std::unordered_map<uint64_t, AF_Order>            pending_;
    std::vector<AF_Position>                          trade_log_;
    uint64_t next_ticket_{1000001};
    mutable std::mt19937 rng_global_{1337};

    void update_equity() const {
        double floating = 0;
        for (auto &[t,p]:positions_) {
            update_position(const_cast<AF_Position&>(p));
            floating += p.profit;
        }
        equity_ = balance_ + floating;
        acct_.equity = equity_;
        acct_.profit = floating;
    }

    void update_position(AF_Position &p) const {
        auto it = price_sim_.find(p.symbol);
        if (it==price_sim_.end()) return;
        double mid = const_cast<PriceSim&>(it->second).next_mid();
        double sp  = default_spread(p.symbol);
        p.current_price= (p.side==AF_DIR_LONG) ? mid-sp/2 : mid+sp/2;
        double raw = (p.side==AF_DIR_LONG)
                     ? (p.current_price-p.open_price)*p.lots*100000.0
                     : (p.open_price-p.current_price)*p.lots*100000.0;
        p.profit = raw + p.commission;
    }
};

/* ── Factory ── */
std::unique_ptr<IBroker> make_paper_broker(double balance) {
    return std::make_unique<PaperBroker>(balance);
}

} /* namespace af */

File: python/algoforge/paper_broker.py
"""PaperBroker — paper trading implementation of IBroker.

Simulates price via a seeded random walk (one per symbol).
Commission: $7/lot charged on open. P&L = (exit − open) × lots × 100 000.
"""
from __future__ import annotations

import math
import random
from typing import Optional

from .types import Bar, Direction
from .broker import (IBroker, OrderType, Tick, AccountInfo,
                     SymbolInfo, Order, Position)

_PAIRS: list[tuple[str, float, float, int]] = [
    ("EURUSD", 1.08500, 0.0080,  1),
    ("GBPUSD", 1.26500, 0.0100,  2),
    ("USDJPY", 149.500, 0.0070,  3),
    ("XAUUSD", 2050.00, 0.0120,  4),
    ("EURJPY", 162.000, 0.0090,  5),
    ("GBPJPY", 188.000, 0.0110,  6),
    ("AUDUSD", 0.65500, 0.0090,  7),
    ("USDCAD", 1.36000, 0.0075,  8),
    ("USDCHF", 0.90500, 0.0070,  9),
    ("EURGBP", 0.85800, 0.0060, 10),
]

_LONG_TYPES  = {OrderType.BUY, OrderType.BUY_LIMIT, OrderType.BUY_STOP}
_MARKET_BUY  = {OrderType.BUY, OrderType.BUY_STOP}
_MARKET_SELL = {OrderType.SELL, OrderType.SELL_STOP}


def _default_spread(sym: str) -> float:
    if "XAU" in sym or "XAG" in sym:
        return 0.25
    if "JPY" in sym:
        return 0.030
    return 0.00012


class _PriceSim:
    def __init__(self, start: float, vol: float, seed: int) -> None:
        self.mid = start
        self._vol = vol
        self._rng = random.Random(seed)

    def next_mid(self) -> float:
        step = self._vol / math.sqrt(24.0 * 60.0)
        self.mid = max(self.mid * (1.0 + self._rng.gauss(0.0, step)), 0.0001)
        return self.mid


class PaperBroker(IBroker):
    def __init__(self, balance: float = 10_000.0) -> None:
        self._balance     = balance
        self._connected   = False
        self._next_ticket = 1_000_001
        self._sims: dict[str, _PriceSim] = {
            sym: _PriceSim(start, vol, seed)
            for sym, start, vol, seed in _PAIRS
        }
        self._positions: dict[int, Position] = {}
        self._rng_global = random.Random(1337)

    # ── Connection ──────────────────────────────────────────────────────────

    def connect(self) -> int:
        self._connected = True
        print(f"[PAPER] Connected | balance={self._balance:.2f}")
        return 0

    def disconnect(self) -> None:
        self._connected = False

    def is_connected(self) -> bool:
        return self._connected

    def broker_name(self) -> str:
        return "PaperSimulator"

    # ── Account ─────────────────────────────────────────────────────────────

    def get_account(self) -> AccountInfo:
        floating = sum(p.profit for p in self._positions.values())
        return AccountInfo(
            balance=self._balance,
            equity=self._balance + floating,
            free_margin=self._balance,
            leverage=100,
            currency="USD",
            login=99999999,
        )

    # ── Market data ──────────────────────────────────────────────────────────

    def get_tick(self, symbol: str) -> Tick:
        sim = self._sims.get(symbol)
        if sim is None:
            raise ValueError(f"Unknown symbol: {symbol}")
        mid = sim.next_mid()
        sp  = _default_spread(symbol)
        return Tick(
            symbol=symbol,
            bid=mid - sp / 2.0,
            ask=mid + sp / 2.0,
            volume=500.0 + self._rng_global.uniform(0, 2000),
        )

    def get_symbol_info(self, symbol: str) -> SymbolInfo:
        if "XAU" in symbol or "XAG" in symbol:
            digits = 2
        elif "JPY" in symbol:
            digits = 3
        else:
            digits = 5
        return SymbolInfo(
            name=symbol, digits=digits,
            point=10.0 ** -digits, contract_size=100_000.0,
            volume_min=0.01, volume_max=100.0, volume_step=0.01,
        )

    def get_bars(self, symbol: str, tf: int, count: int) -> list[Bar]:
        sim = self._sims.get(symbol)
        if sim is None:
            raise ValueError(f"Unknown symbol: {symbol}")
        rng = random.Random(42 ^ tf)
        vol_per_bar = 0.0080 / math.sqrt(86400.0 / tf) / math.sqrt(252.0)
        price = sim.mid
        bars: list[Bar] = []
        for _ in range(count):
            o = price
            h = l = o
            for _ in range(4):
                p = o * (1.0 + rng.gauss(0.0, vol_per_bar))
                h = max(h, p)
                l = min(l, p)
            c = o * (1.0 + rng.gauss(0.0, vol_per_bar))
            h = max(h, c)
            l = min(l, c)
            bars.append(Bar(timestamp=0, open=o, high=h, low=l, close=c,
                            volume=500.0 + rng.uniform(0, 4500),
                            spread=_default_spread(symbol)))
            price = c
        return bars

    # ── Trading ──────────────────────────────────────────────────────────────

    def place_order(self, symbol: str, order_type: OrderType,
                    lots: float, price: float = 0.0,
                    sl: float = 0.0, tp: float = 0.0,
                    magic: int = 0, comment: str = "") -> Order:
        if lots < 0.01:
            raise ValueError("lots must be >= 0.01")
        sim = self._sims.get(symbol)
        if sim is None:
            raise ValueError(f"Unknown symbol: {symbol}")

        mid = sim.next_mid()
        sp  = _default_spread(symbol)
        bid, ask = mid - sp / 2.0, mid + sp / 2.0

        if order_type in _MARKET_BUY:
            fill_p = ask
        elif order_type in _MARKET_SELL:
            fill_p = bid
        elif order_type == OrderType.BUY_LIMIT:
            fill_p = min(price, ask)
        else:
            fill_p = max(price, bid)

        is_long    = order_type in _LONG_TYPES
        commission = 7.0 * lots
        ticket     = self._next_ticket
        self._next_ticket += 1

        pos = Position(
            ticket=ticket, symbol=symbol,
            side=Direction.LONG if is_long else Direction.SHORT,
            lots=lots, open_price=fill_p, current_price=fill_p,
            sl=sl, tp=tp, profit=0.0, commission=-commission,
            magic=magic, comment=comment,
        )
        self._positions[ticket] = pos
        self._balance -= commission

        print(f"[PAPER] {'BUY' if is_long else 'SELL'} {lots:.2f} {symbol} @ {fill_p:.5f}"
              f" | SL={sl:.5f} TP={tp:.5f} | comm=-{commission:.2f}")
        return Order(ticket=ticket, symbol=symbol, type=order_type,
                     lots=lots, fill_price=fill_p, sl=sl, tp=tp, magic=magic)

    def close_position(self, ticket: int, lots: float = 0.0) -> int:
        pos = self._positions.get(ticket)
        if pos is None:
            return -6  # AF_ERR_POSITION_NOT_FOUND
        sim   = self._sims.get(pos.symbol)
        mid   = sim.next_mid() if sim else pos.open_price
        sp    = _default_spread(pos.symbol)
        exit_p = mid - sp / 2.0 if pos.side == Direction.LONG else mid + sp / 2.0
        close_lots = lots if (0 < lots < pos.lots) else pos.lots
        raw_pnl = ((exit_p - pos.open_price) if pos.side == Direction.LONG
                   else (pos.open_price - exit_p)) * close_lots * 100_000.0
        self._balance += raw_pnl
        print(f"[PAPER] CLOSE {pos.symbol} "
              f"{'LONG' if pos.side == Direction.LONG else 'SHORT'} "
              f"{close_lots:.2f} @ {exit_p:.5f} | P&L={raw_pnl:.2f}")
        if close_lots >= pos.lots:
            del self._positions[ticket]
        else:
            pos.lots -= close_lots
        return 0

    def modify_position(self, ticket: int, sl: float, tp: float) -> int:
        pos = self._positions.get(ticket)
        if pos is None:
            return -6
        pos.sl = sl
        pos.tp = tp
        return 0

    def cancel_order(self, ticket: int) -> int:
        return 0

    # ── Queries ──────────────────────────────────────────────────────────────

    def get_positions(self) -> list[Position]:
        return list(self._positions.values())

    def get_orders(self) -> list[Order]:
        return []

    def get_position(self, ticket: int) -> Optional[Position]:
        return self._positions.get(ticket)

    # ── Simulation helper ────────────────────────────────────────────────────

    def poll_sl_tp(self) -> None:
        to_close: list[int] = []
        for ticket, pos in self._positions.items():
            sim = self._sims.get(pos.symbol)
            if not sim:
                continue
            mid = sim.next_mid()
            sp  = _default_spread(pos.symbol)
            bid, ask = mid - sp / 2.0, mid + sp / 2.0
            pos.current_price = bid if pos.side == Direction.LONG else ask
            hit = False
            if pos.sl > 0:
                hit = (pos.side == Direction.LONG  and bid <= pos.sl) or \
                      (pos.side == Direction.SHORT and ask >= pos.sl)
            if not hit and pos.tp > 0:
                hit = (pos.side == Direction.LONG  and bid >= pos.tp) or \
                      (pos.side == Direction.SHORT and ask <= pos.tp)
            if hit:
                to_close.append(ticket)
        for t in to_close:
            self.close_position(t)


def make_paper_broker(balance: float = 10_000.0) -> PaperBroker:
    return PaperBroker(balance)
